godot-core/src/builtin/quaternion.rs

```diff
@@ -130,70 +138,115 @@ impl Quaternion {
-    // TODO: Implement godot's function in Rust
-    /*
-        pub fn spherical_cubic_interpolate_in_time(
-            self,
-            b: Self,
-            pre_a: Self,
-            post_b: Self,
-            weight: real,
-            b_t: real,
-            pre_a_t: real,
-            post_b_t: real,
-        ) -> Self {
-        }
-    */
+        let interpolated = self.as_inner().slerpni(to, weight.as_f64());
+
+        let error = interpolated.is_default() && (!self.is_normalized() || !to.is_normalized());
+        assert!(!error, "Slerpni requires normalized quaternions");
+
+        interpolated
+    }
+
+    /// Performs spherical cubic interpolation.
+    ///
+    /// # Panics
+    /// If any quaternions are not normalized.
+    pub fn spherical_cubic_interpolate(
+        self,
+        b: Self,
+        pre_a: Self,
+        post_b: Self,
+        weight: real,
+    ) -> Self {
+        let interpolated =
+            self.as_inner()
+                .spherical_cubic_interpolate(b, pre_a, post_b, weight.as_f64());
+        // Godot returns default if you give it quaternions that are not normalized. This means we can check for default
+        // Then check if we should panic.
+        let error = interpolated.is_default() && !self.are_normalized(b, pre_a, post_b);
+        assert!(
+            !error,
+            "Spherical cubic interpolation requires normalized quaternions"
+        );
+
+        interpolated
+    }
+
+    /// Performs spherical cubic interpolation in time.
+    ///
+    /// # Panics
+    /// If any quaternions are not normalized.
+    #[allow(clippy::too_many_arguments)]
+    pub fn spherical_cubic_interpolate_in_time(
+        self,
+        b: Self,
+        pre_a: Self,
+        post_b: Self,
+        weight: real,
+        b_t: real,
+        pre_a_t: real,
+        post_b_t: real,
+    ) -> Self {
+        let interpolated = self.as_inner().spherical_cubic_interpolate_in_time(
+            b,
+            pre_a,
+            post_b,
+            weight.as_f64(),
+            b_t.as_f64(),
+            pre_a_t.as_f64(),
+            post_b_t.as_f64(),
+        );
+        // Godot returns default if you give it quaternions that are not normalized. This means we can check for default
+        // Then check if we should panic.
+        let error = interpolated.is_default() && !self.are_normalized(b, pre_a, post_b);
+        assert!(
+            !error,
+            "Spherical cubic interpolation in time requires normalized quaternions"
+        );
+        interpolated
+    }
 
     #[doc(hidden)]
     pub fn as_inner(&self) -> inner::InnerQuaternion {
         inner::InnerQuaternion::from_outer(self)
     }
+
+    #[doc(hidden)]
+    pub fn is_default(&self) -> bool {
+        self.x == 0.0 && self.y == 0.0 && self.z == 0.0 && self.w == 1.0
+    }
+
+    #[doc(hidden)]
+    pub fn are_normalized(&self, a: Self, b: Self, c: Self) -> bool {
```

Comment:
could take a `vec![]` opted for YAGNI

Reply:
In general, YAGNI is king 👑 

_However_, now that I see the number of times you use assertions of the pattern
 `x.is_default() && <varying number of args>.is_normalized()`, you could really think about something like:
```rs
fn ensure_normalized(&self, quats: &[&Quaternion]) {
    ... // assert etc
}
```